### PR TITLE
Ported Jump-Point-Search pathing from TG

### PR DESCRIPTION
**Pull request by DamianX**
This is better than the current AStar system because I can get it to work.
Ported from https://github.com/tgstation/tgstation/pull/56780 with permission from the author.

No player-visible changes. I only implemented it for the pathing debug robot for now.

### CODE REVIEW

File: code/__HELPERS/heap.dm [hunk 1]
```diff
@@ -0,0 +1,80 @@
+//////////////////////
+//datum/heap object
```
Comment (Exxion) [inline code/__HELPERS/heap.dm:2]: Oh hey you can use this for the timer subsystem too

**Comment by SonixApache** [comment]
might be an idea to add test shit like this to xenoarch finds - low availability, free crowdsource testing, fun for players

**Comment by DeityLink** [comment]
Oh for the love of god can this fix bots? I'm tired of Mulebot getting stuck in pathfinding hell and never moving again for whatever reason.

File: __DEFINES/callback.dm [hunk 2]
```diff
@@ -1,2 +1,2 @@
 // Set a callback's thing_to_call to GLOBAL_PROC in order to call a global proc.
-#define GLOBAL_PROC "fuck you duny"
```
Comment (Kurfursten) [inline __DEFINES/callback.dm:2]: hmm

Comment (DamianX) [reply __DEFINES/callback.dm:2]: I added it, I taketh away.

Comment (Exxion) [reply __DEFINES/callback.dm:2]: This change was part of the license agreement for porting this code. Actually the entirety of the license agreement.

**Pull request merged**